independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java

```diff
@@ -625,22 +625,69 @@ public List<ClassPathElement> getElementsWithResource(String name) {
     public List<ClassPathElement> getElementsWithResource(String name, boolean localOnly) {
         ensureOpen(name);
 
-        boolean parentFirst = parentFirst(name, getClassPathResourceIndex());
+        final boolean parentFirst = parentFirst(name, getClassPathResourceIndex());
 
-        List<ClassPathElement> ret = new ArrayList<>();
+        List<ClassPathElement> result = List.of();
 
-        if (parentFirst && !localOnly && parent instanceof QuarkusClassLoader) {
-            ret.addAll(((QuarkusClassLoader) parent).getElementsWithResource(name));
+        if (parentFirst && !localOnly && parent instanceof QuarkusClassLoader parentQcl) {
+            result = parentQcl.getElementsWithResource(name);
         }
 
-        List<ClassPathElement> classPathElements = getClassPathResourceIndex().getClassPathElements(name);
-        ret.addAll(classPathElements);
+        result = unify(result, getClassPathResourceIndex().getClassPathElements(name));
 
-        if (!parentFirst && !localOnly && parent instanceof QuarkusClassLoader) {
-            ret.addAll(((QuarkusClassLoader) parent).getElementsWithResource(name));
+        if (!parentFirst && !localOnly && parent instanceof QuarkusClassLoader parentQcl) {
+            result = unify(result, parentQcl.getElementsWithResource(name));
         }
 
-        return ret;
+        return result;
+    }
+
+    /**
+     * Returns a list containing elements from two lists eliminating duplicates. Elements from the first list
+     * will appear in the result before elements from the second list.
+     * <p>
+     * The current implementation assumes that none of the lists contains duplicates on their own but some elements
+     * may be present in both lists.
+     *
+     * @param list1 first list
+     * @param list2 second list
+     * @return resulting list
+     */
+    private static List<ClassPathElement> unify(List<ClassPathElement> list1, List<ClassPathElement> list2) {
```

Comment:
Minor nitpick: I would prefer if this method was called `deduplicate`

Reply:
`joinAndDedupe`? :)

Reply:
👌